cumulusci/tasks/bulkdata/load.py

```diff
@@ -142,42 +144,58 @@ def _execute_step(
 
         mapping["oid_as_pk"] = bool(mapping.get("fields", {}).get("Id"))
 
-        bulk_mode = mapping.get("bulk_mode") or self.bulk_mode or "Parallel"
-
-        step = BulkApiDmlOperation(
-            sobject=mapping["sf_object"],
-            operation=(
-                DataOperationType.INSERT
-                if mapping.get("action") == "insert"
-                else DataOperationType.UPDATE
-            ),
-            api_options={"bulk_mode": bulk_mode},
-            context=self,
-            fields=self._get_columns(mapping),
-        )
-
         local_ids = []
+        query = self._query_db(mapping)
+
+        api = mapping.api
+        if api is DataApi.SMART:
+            record_count = query.count()
+            api = DataApi.BULK if record_count > 10000 else DataApi.REST
+
+        if api is DataApi.BULK:
+            bulk_mode = mapping.get("bulk_mode") or self.bulk_mode or "Parallel"
+
+            step = BulkApiDmlOperation(
+                sobject=mapping["sf_object"],
+                operation=(
+                    DataOperationType.INSERT
+                    if mapping.get("action") == "insert"
+                    else DataOperationType.UPDATE
+                ),
+                api_options={"bulk_mode": bulk_mode},
+                context=self,
+                fields=self._get_columns(mapping),
+            )
+        elif api is DataApi.REST:
+            step = RestApiDmlOperation(
+                sobject=mapping["sf_object"],
+                operation=(
+                    DataOperationType.INSERT
+                    if mapping.get("action") == "insert"
+                    else DataOperationType.UPDATE
+                ),
+                api_options={"batch_size": mapping.batch_size},
+                context=self,
+                fields=self._get_columns(mapping),
+            )
+
         step.start()
-        step.load_records(self._stream_queried_data(mapping, local_ids))
+        step.load_records(self._stream_queried_data(mapping, local_ids, query))
         step.end()
 
         if step.job_result.status is not DataOperationStatus.JOB_FAILURE:
             self._process_job_results(mapping, step, local_ids)
 
         return step.job_result
 
-    def _stream_queried_data(self, mapping, local_ids):
+    def _stream_queried_data(self, mapping, local_ids, query):
         """Get data from the local db"""
 
         statics = self._get_statics(mapping)
-        query = self._query_db(mapping)
 
         total_rows = 0
 
-        # 10,000 is the maximum Bulk API size. Clamping the yield from the query ensures we do not
-        # create more Bulk API batches than expected, regardless of batch size, while capping
-        # memory usage.
-        for row in query.yield_per(10000):
```

Comment:
I know the bulk API operations now have improved logic for creating batches that fit within limits, but this is still useful for making sure we don't try to load all rows from the SQL db into memory at once, which can be a problem when processing millions of records.

Reply:
I've restored this element.